pkg/trie/triedb/codec/decode.go

```diff
@@ -0,0 +1,162 @@
+
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+)
+
+var (
+	ErrDecodeHashedStorageValue  = errors.New("cannot decode hashed storage value")
+	ErrDecodeHashedValueTooShort = errors.New("hashed storage value too short")
+	ErrReadChildrenBitmap        = errors.New("cannot read children bitmap")
+	// ErrDecodeChildHash is defined since no sentinel error is defined
+	// in the scale package.
+	ErrDecodeChildHash = errors.New("cannot decode child hash")
+	// ErrDecodeStorageValue is defined since no sentinel error is defined
+	// in the scale package.
+	ErrDecodeStorageValue = errors.New("cannot decode storage value")
+)
+
+const hashLength = common.HashLength
+
+// Decode decodes a node from a reader.
+// The encoding format is documented in the README.md
+// of this package, and specified in the Polkadot spec at
+// https://spec.polkadot.network/chap-state#defn-node-header
+func Decode(reader io.Reader) (n Node, err error) {
+	variant, partialKeyLength, err := decodeHeader(reader)
+	if err != nil {
+		return nil, fmt.Errorf("decoding header: %w", err)
+	}
+
+	switch variant {
+	case emptyVariant:
+		return Empty{}, nil
+	case leafVariant, leafWithHashedValueVariant:
+		n, err = decodeLeaf(reader, variant, partialKeyLength)
+		if err != nil {
+			return nil, fmt.Errorf("cannot decode leaf: %w", err)
+		}
+		return n, nil
+	case branchVariant, branchWithValueVariant, branchWithHashedValueVariant:
+		n, err = decodeBranch(reader, variant, partialKeyLength)
+		if err != nil {
+			return nil, fmt.Errorf("cannot decode branch: %w", err)
+		}
+		return n, nil
+	default:
+		// this is a programming error, an unknown node variant should be caught by decodeHeader.
+		panic(fmt.Sprintf("not implemented for node variant %08b", variant))
+	}
+}
+
+// decodeBranch reads from a reader and decodes to a node branch.
+// Note that we are not decoding the children nodes.
+func decodeBranch(reader io.Reader, variant variant, partialKeyLength uint16) (
+	node Branch, err error) {
+	node = Branch{}
+
+	node.PartialKey, err = decodeKey(reader, partialKeyLength)
+	if err != nil {
+		return Branch{}, fmt.Errorf("cannot decode key: %w", err)
+	}
+
+	childrenBitmap := make([]byte, 2)
+	_, err = reader.Read(childrenBitmap)
+	if err != nil {
+		return Branch{}, fmt.Errorf("%w: %s", ErrReadChildrenBitmap, err)
+	}
+
+	sd := scale.NewDecoder(reader)
+
+	switch variant {
+	case branchWithValueVariant:
+		valueBytes := make([]byte, 0)
+		err := sd.Decode(&valueBytes)
+
+		node.Value = NewInlineValue(valueBytes)
+		if err != nil {
+			return Branch{}, fmt.Errorf("%w: %s", ErrDecodeStorageValue, err)
+		}
+	case branchWithHashedValueVariant:
+		hashedValue, err := decodeHashedValue(reader)
+		if err != nil {
+			return Branch{}, err
+		}
+		node.Value = NewHashedValue(hashedValue)
+	default:
+		// Do nothing, branch without value
+	}
+
+	for i := 0; i < ChildrenCapacity; i++ {
+		if (childrenBitmap[i/8]>>(i%8))&1 != 1 {
+			continue
+		}
```

Comment:
I think that this is a little bit hard to understand, I would like to suggest another approach:

```go 
// instead of dealing with a 2 items byte slice like `childrenBitmap := make([]byte, 2)` we deal with a // uint16 (which is also represented by a 2 byte slice)

var childrenBitmap uint16
err := binary.Read(reader, binary.LittleEndian, &bitmap) // will read 2 byte from the reader
if err != nil { ... }

// then you can go over the 16 positions checking each bit in the uint16 value
// but on every loop you shift childrenBitmap 1 to the left
for i := 0; i < ChildrenCapacity; i++{
   // this check is optional since we cover all bits (16) 
   // available in uint16 type
    if childrenBitmap == 0 {
        break;
     }
     
     // just check if we have a child in 
     // the current bit, if so, just decode
      if (childrenBitmap & 1) == 1 {
            err := decodeBranchChild()
            if err != nil {...}
       }
      
      // otherwise advance to check next bit in the bitmap
       childrenBitmap >>= 1
}

func decodeBranchChild() {
                 var hash []byte
		err := sd.Decode(&hash)
		if err != nil {
			return Branch{}, fmt.Errorf("%w: at index %d: %s",
				ErrDecodeChildHash, i, err)
		}

		if len(hash) < hashLength {
			node.Children[i] = NewInlineNode(hash)
		} else {
			node.Children[i] = NewHashedNode(hash)
		}
}
```

Reply:
mmm, I agree that it's complicated to understand but I think we can solve it using a mix of both approaches.

1. We can parse the value into an `uint16` to not use an array of two elements
2. We can simplify the check to `if (childrenBitmap>>i)&1 != 1`

I think it's more understandable now with less changes

Reply:
Agreed!